pika/connection.py

```diff
@@ -1667,40 +1714,30 @@ def _read_frame(self):
         """
         return frame.decode_frame(self._frame_buffer)
 
-    def _reject_out_of_band_delivery(self, channel_number, delivery_tag):
```

Comment:
@gmr, regarding
> Just to reiterate, I think there is a case where you could have a pile of frames on the stack that are unprocessed when a Channel.Close is emitted. My thinking is that issuing the close should be similar to Basic.Cancel. I guess one could make the argument that both should not be respected until their respective Ok frames are received though.

In response to your other comment, I provided a detailed explanation of why that should not be possible, and also described a race condition bug that I fixed in this PR that used to allow for that protocol breach in pika. As a quick summary for completeness here, pika will not remove the channel number from `Connection._channels` until it receives the proper acknowledgement from the broker. Therefore, it would be impossible to receive frames for that channel after it's removed from `Connection._channels` - the broker will not send any more messages on the channel after sending `Basic.CloseOk` or `Basic.Close` to the client on that channel. It would be a violation of the AMQP protocol if it sent more frames after sending `Basic.CloseOk` or `Basic.Close` on the same channel.

Regarding your comparison with `Basic.Cancel`, there is one important difference between it and `Channel.Close`: `Channel.Close` is only supported in synchronous mode - i.e., the broker must respond with `Channel.CloseOk` according to AMQ 0.9.1. On the other hand: although `Basic.Cancel` is defined a synchronous, it may be used asynchronously by setting `nowait=True` (unfortunate in my opinion), therefore eliminating `Basic.CancelOk`, so there is no simple way to tell when the Consumer is fully closed. As the result, `pika.Channel` ends up leaking some data structures (in `self._cancelled`, I think). In any case, `Channel.Close` flow doesn't have the problem that `Basic.Cancel` has. (with Consumers, we could probably safely clean up the leaked consumer-related object(s) after the next syncronous `_rpc` request completes, but that's another story for another day :) )